cpp/density/GaussianDensity.cc

```diff
@@ -31,94 +31,107 @@ vec3<unsigned int> GaussianDensity::getWidth()
 
             // In 2D, only loop over the z=0 plane
             if (m_box.is2D())
             {
                 bin_z = 0;
             }
 
+            // Reject bins that are outside the box in aperiodic directions
             // Only evaluate over bins that are within the cutoff
             for (int k = bin_z - bin_cut_z; k <= bin_z + bin_cut_z; k++)
             {
-                const float dz = float((grid_size_z * k + grid_size_z / 2.0f) - point.z - lz / 2.0f);
+                if (!periodic.z && (k < 0 || k >= int(m_width.z)))
+                {
+                    continue;
+                }
+                const float dz = float((grid_size_z * k + grid_size_z / 2.0f) - point.z - Lz / 2.0f);
 
                 for (int j = bin_y - bin_cut_y; j <= bin_y + bin_cut_y; j++)
                 {
-                    const float dy = float((grid_size_y * j + grid_size_y / 2.0f) - point.y - ly / 2.0f);
+                    if (!periodic.y && (j < 0 || j >= int(m_width.y)))
+                    {
+                        continue;
+                    }
+                    const float dy = float((grid_size_y * j + grid_size_y / 2.0f) - point.y - Ly / 2.0f);
 
                     for (int i = bin_x - bin_cut_x; i <= bin_x + bin_cut_x; i++)
                     {
+                        if (!periodic.x && (i < 0 || i >= int(m_width.x)))
+                        {
+                            continue;
+                        }
+                        const float dx = float((grid_size_x * i + grid_size_x / 2.0f) - point.x - Lx / 2.0f);
+
                         // Calculate the distance from the particle to the grid cell
-                        const float dx = float((grid_size_x * i + grid_size_x / 2.0f) - point.x - lx / 2.0f);
-                        vec3<float> delta = m_box.wrap(vec3<float>(dx, dy, dz));
+                        const vec3<float> delta = m_box.wrap(vec3<float>(dx, dy, dz));
 
                         const float r_sq = dot(delta, delta);
-                        const float r_sqrt = std::sqrt(r_sq);
 
                         // Check to see if this distance is within the specified r_max
-                        if (r_sqrt < m_r_max)
+                        if (r_sq < r_max_sq)
                         {
-                            // Evaluate the gaussian ...
-                            const float x_gaussian
-                                = A * std::exp((-1.0f) * (delta.x * delta.x) / (2.0f * sigmasq));
-                            const float y_gaussian
-                                = A * std::exp((-1.0f) * (delta.y * delta.y) / (2.0f * sigmasq));
-                            const float z_gaussian
-                                = A * std::exp((-1.0f) * (delta.z * delta.z) / (2.0f * sigmasq));
+                            // Evaluate the gaussian
+                            const float gaussian = normalization * std::exp(-r_sq / (2 * sigmasq));
```

Comment:
The variable `sigmasq` is a float, so what is the need for casting here? Also, isn't casting in cpp done as `(float) 2.0`, not `float(2.0)`?

Reply:
You're mostly right about these two things. The `(float) 2.0` style is actually C-style casting, not C++ style. The function style cast `float(2.0)` was added in C++, but it behaves exactly like C-style casts except in cases where there are multiple arguments, in which case it is capable of calling more complex constructors.

The preferred method of casting in C++ is actually by explicitly specifying the type of the cast, in this case `static_cast` (https://en.cppreference.com/w/cpp/language/static_cast). 
C-style casts are pretty unsafe in general since they will even try to cast things that aren't safe to cast to each other. However, in cases like this where the object is a float literal I don't personally worry about that. If we wanted to *strictly* adhere to best practices we would want to go through the entire code base and replace most such casts with the `static_cast` variant, but there's not really a good reason to.

You're also right that casting isn't strictly necessary here. The reason I made the change is in anticipation of a future release of freud that supports variable/mixed precision (see #464). The easiest way to do that is something like what HOOMD does, where there's a global typedef of the desired type (`Scalar` in HOOMD) and all literals are typecast throughout the code base. This of course relies on compilers to optimize away the type conversion since otherwise you don't really get the same performance tradeoffs, but that's a pretty safe assumption in such simple cases AFAIK.